File: lld_game_design_primer/snake_ladder/Subproblem/SnakeAndLadder.py

import math


class SnakeAndLadder:

    def __init__(self):
        self.diceRollSides = 6

    def snakesAndLadders(self, board):
        board_size = len(board)
        j = 1
        while j <= board_size * board_size:
            coord = self._getCoordinate(j, board_size)
            if board[coord[0]][coord[1]] == -1:
                board[coord[0]][coord[1]] = j
            j += 1
        if (math.fmod(board_size, 2) == 0 and board[0][0] != board_size * board_size) or (
                math.fmod(board_size, 2) != 0 and board[0][board_size - 1] != board_size * board_size):
            return -1
        source = 1
        destination = board_size * board_size
        bfsQueue = []
        visited = set()
        bfsQueue.append(source)
        numberOfEdgesFromSource = [0 for _ in range(board_size * board_size)]
        count = 0
        while not len(bfsQueue) == 0:
            count += 1
            queueSize = len(bfsQueue)
            for i in range(0, queueSize):
                curr = bfsQueue.pop()
                visited.add(curr)
                j = 1
                while j <= self.diceRollSides:
                    coordinate = self._getCoordinate(curr + j, board_size)
                    cellValue = board[coordinate[0]][coordinate[1]]
                    if cellValue == destination:
                        return count
                    if cellValue not in visited:
                        bfsQueue.append(cellValue)
                    j += 1
        return -1

    @staticmethod
    def _getCoordinate(num, dimension):
        equivalentZeroBasedIndex = num - 1
        evenRow = math.fmod((math.trunc((num - 1) / float(dimension))), 2) == 0
        row = (dimension - 1) - (math.trunc((num - 1) / float(dimension)))
        col = math.fmod((num - 1), dimension)
        if not evenRow:
            col = dimension - 1 - col
        return [row, col]

File: lld_game_design_primer/snake_ladder/Subproblem/SnakeAndLadder.java
package lld_game_design_primer.snake_ladder.Subproblem;

import java.util.ArrayDeque;
import java.util.HashSet;
import java.util.Queue;
import java.util.Set;

class SnakeAndLadder {
    int diceRollSides = 6; // number of sides the dice has
    public int snakesAndLadders(int[][] board) {
        int len = board.length;
        for (int j = 1; j <= len * len; j++) {
            int[] coord = getCoordinate(j, len);
            if (board[coord[0]][coord[1]] == -1) {
                board[coord[0]][coord[1]] = j;
            }
        }
        
        // the only time that there would be no solution is if the square (n ^ 2)
        // has a snake mouth, since every time you reach here you go down to some other square
        // so there is no way you can stay at square (n ^ 2)
        if ((len % 2 == 0 && board[0][0] != len * len) 
                    || (len % 2 != 0 && board[0][len - 1] != len * len)) {
            return -1;
        }
        
        int source = 1;
        int destination = len * len;
        
        Queue<Integer> bfsQueue = new ArrayDeque<>();
        Set<Integer> visited = new HashSet<>(); // hashset offers efficient search
        bfsQueue.add(source);
        
        int count = 0;
        // BFS
        while (!bfsQueue.isEmpty()) {
            count++;
            int queueSize = bfsQueue.size();
            for (int i = 0; i < queueSize; i++) {
                int curr = bfsQueue.poll();
                visited.add(curr);
                for (int j = 1; j <= diceRollSides; j++) {
                    int[] coordinate = getCoordinate(curr + j, len);
                    int cellValue = board[coordinate[0]][coordinate[1]];
                    if (cellValue == destination) {
                        return count;
                    }
                    if (!visited.contains(cellValue)) {
                        bfsQueue.add(cellValue);
                    }
                }
            }
        }
        return -1;
    }
    
    private int[] getCoordinate(int num, int dimension) {
        boolean evenRow = ((num - 1) / dimension) % 2 == 0; // zero based -> row index starts from 0 which is even
                                // all even rows have numbers in ascending order from left to right.
                                // this indexing is done brom bottom to top which 
                                // is reverse of the index order in given matrix/board.
        int row = (dimension - 1) - ((num - 1) / dimension);
        int col = (num - 1) % dimension;
        if (!evenRow) {
            col = dimension - 1 - col;
        }
        return new int[] {row, col};
    }
}